engines/cryomni3d/dialogs_manager.cpp

```diff
@@ -0,0 +1,558 @@
+ *
+ */
+
+#include "cryomni3d/dialogs_manager.h"
+
+#include "common/debug.h"
+#include "common/file.h"
+
+namespace CryOmni3D {
+
+DialogsManager::~DialogsManager() {
+	delete[] _gtoBuffer;
+}
+
+void DialogsManager::loadGTO(const Common::String &gtoFileName) {
+	Common::File gtoFile;
+	if (!gtoFile.open(gtoFileName)) {
+		error("Can't open GTO file '%s'", gtoFileName.c_str());
+	}
+
+	_labels.clear();
+	_gtoEnd = nullptr;
+	delete[] _gtoBuffer;
+	_gtoBuffer = nullptr;
+
+	unsigned int gtoSize = gtoFile.size();
+	_gtoBuffer = new char[gtoSize];
+	gtoFile.read(_gtoBuffer, gtoSize);
+	gtoFile.close();
+
+	_gtoEnd = _gtoBuffer + gtoSize;
+
+	populateLabels();
+}
+
+void DialogsManager::populateLabels() {
+	/* Get labels count and populate the labels array */
+	unsigned int numLabels;
+	const char *labelsP = strstr(_gtoBuffer, "LABELS=");
+	if (labelsP) {
+		labelsP += sizeof("LABELS=") - 1;
+		for (; *labelsP == ' '; labelsP++) { }
+		numLabels = atoi(labelsP);
+	} else {
+		numLabels = 0;
+	}
+
+	for (const char *labelP = _gtoBuffer; labelP != nullptr; labelP = nextLine(labelP)) {
+		if (*labelP == ':') {
+			/* Line starting with ':', it's a label */
+			_labels.push_back(nextChar(labelP));
+		}
+	}
+
+	if (_labels.size() != numLabels) {
+		error("Bad labels number in GTO");
+	}
+}
+
+const char *DialogsManager::findLabel(const char *label, const char **realLabel) const {
+	unsigned int labelLen = 0;
+	/* Truncate input label */
+	for (const char *labelP = label;
+	        *labelP != '\0' &&
+	        *labelP != ' ' &&
+	        *labelP != '.' &&
+	        *labelP != '\r'; labelP++, labelLen++) { }
+
+	Common::Array<const char *>::const_iterator labelsIt;
+	for (labelsIt = _labels.begin(); labelsIt != _labels.end(); labelsIt++) {
+		if (!strncmp(*labelsIt, label, labelLen)) {
+			break;
+		}
+	}
+
+	if (labelsIt == _labels.end()) {
+		error("Label not found");
+	}
+
+	if (realLabel) {
+		*realLabel = *labelsIt;
+	}
+	return nextLine(*labelsIt);
+}
+
+Common::String DialogsManager::getLabelSound(const char *label) const {
+	/* Remove starting : if any */
+	if (*label == ':') {
+		label++;
+	}
+
+	const char *labelEnd;
+	for (labelEnd = label; *labelEnd >= '0' && *labelEnd <= 'Z'; labelEnd++) { }
+
+	return Common::String(label, labelEnd);
+}
+
+const char *DialogsManager::findSequence(const char *sequence) const {
+	unsigned int sequenceLen = strlen(sequence);
+
+	const char *lineP;
+	for (lineP = _gtoBuffer; lineP != nullptr; lineP = nextLine(lineP)) {
+		if (!strncmp(lineP, sequence, sequenceLen)) {
+			/* Line starting with the sequence name */
+			break;
+		}
+	}
+
+	if (!lineP) {
+		return nullptr;
+	}
+
+	/* Find next label */
+	for (; lineP != nullptr && *lineP != ':'; lineP = nextLine(lineP)) { }
+
+	/* Return the label name without it's ':' */
+	return nextChar(lineP);
+}
+
+Common::String DialogsManager::findVideo(const char *data) const {
+	data = previousMatch(data, ".FLC");
+	if (data == nullptr) {
+		return nullptr;
```

Comment:
I get an error here: no conversion from nullptr to Common::String
So I changed this to:
```
		return Common::String();
```

Reply:
I don't even know how GCC let that pass without any error or warning.
I will try to address your comments this evening.